### ISSUE TITLE
Question: providers.Configuration from *.ini

### ISSUE THREAD

**Issue by thegrymek** [ets-labs/python-dependency-injector#192]
Hi,

I use this module daily at work and for me fits perfectly! Thanks so much for your work! 

Question:
Is there any way to load ini file directly to `dependency_injector.providers.Configuration`?
Currently I solve this by implementing in `Core` my configuration loader and then override config values.
Other way is to use directly python `ConfigParser`, load `ini` file and then override `Core.config`. I don't want to use directly `ConfigParser` because I had too much customized this class. 

So there is another question:
Is there any awy to use customized `ConfigParser` to paste parameters from `ini` to `Core.config`?

best regards!

**Comment by rmk135** [comment]
Hi @thegrymek ,

Thanks a lot for your feedback, really happy that Dependency Injector works for you.

Regarding your question:

> Is there any awy to use customized ConfigParser to paste parameters from ini to Core.config?

- Unfortunately, currently it is not possible.

I did not include too much of details about where and how configuration is stored and loaded. There are quite a lot of options. From the other point of view, may be it's not a bad idea to extend Dependency Injector with this functionality. Will think of it...

Thanks,
Roman

**Comment by thegrymek** [comment]
I faced with problem to load multiple configuration files from only one project.

One config ini file for loggers, another one for hardcoded constant values, and one more for options like ip, passwords for databases.

My proposal is to extend `provides.Configuration` for methods: `.from_ini(filepath)`.

**Comment by LastStarDust** [comment]
@thegrymek @rmk135 

I am posting here because my question is closely related to this issue. In case I am not getting any answer, I might open a new issue (or if there is a dedicated forum let me know).

I would like to use the dependency-injector module to provide configuration variables to all the modules of my project. I would like to read the configuration from a .ini file using ConfigParser just like the OP. The OP described very briefly how he is achieving the goal. Since I have never used the dependency-injector module I am not sure what he means ... **is my interpretation below correct?**

```python
import configparser

import dependency_injector.containers
import dependency_injector.providers


class MyConfigParser:
    def __init__(self):
        self._configuration = configparser.ConfigParser()
        # READ CONFIGURATINO FILE HERE
        self._configuration.add_section('a')  # random stuff
        self._configuration.set('a', 'foo1', 'bar1')  # random stuff
        self._configuration.add_section('b')  # random stuff
        self._configuration.set('b', 'foo2', 'bar2')  # random stuff

    @property
    def a(self):
        return dict(self._configuration.items('a'))

    @property
    def b(self):
        return dict(self._configuration.items('b'))


class Configuration(dependency_injector.containers.DeclarativeContainer):
    parser = MyConfigParser()
    a = dependency_injector.providers.Configuration('a')
    a.override(parser.a)
    b = dependency_injector.providers.Configuration('b')
    b.override(parser.b)


if __name__ == "__main__":
    # USE VALUE FROM CONFIGURATION FILE
    print(Configuration.a.foo1())
    print(Configuration.b.foo2())

    # OVERRIDE CONFIGURATION (FOR EXAMPLE DURING UNIT TESTS)
    Configuration.a.override({'foo1': 'babar1'})
    Configuration.b.override({'foo2': 'babar2'})
    print(Configuration.a.foo1())
    print(Configuration.b.foo2())
```

My other question is: in the implementation above, the methods foo1() and foo2() are dynamically accessed. This would make it difficult to detect bugs and typos at an early stage and make IDE autocompletion impossible. **So would it be possible to have foo1() and foo2() and their values statically typed?**

**Comment by thegrymek** [comment]
Hi,

I dont remember exactly how I did it. Try with something like that. I used example from here:

* https://github.com/ets-labs/python-dependency-injector/blob/master/examples/miniapps/services_v1/containers.py

```python
```

**Comment by thegrymek** [comment]
But better solution is to load configuration seperately.  Bear in mind that this package should only allows to configure in easy way your objects. If you have ini configuration, load it else where. 

Look at this example:
* https://github.com/ets-labs/python-dependency-injector/blob/master/examples/miniapps/services_v1/run.py

For this dictionary from example:

```python
{'database': {'dsn': ':memory:'},
                          'aws': {'access_key_id': 'KEY',
                                  'secret_access_key': 'SECRET'},
                          'auth': {'token_ttl': 3600}}
```

You should have some methods or lines that loads all configs and final result convert to dict and paste it to `config.override(your_config)`

**Comment by LastStarDust** [comment]
@thegrymek 
Thank you very much for the quick reply.

> But better solution is to load configuration seperately. Bear in mind that this package should only allows to configure in easy way your objects. If you have ini configuration, load it else where.
> 
> Look at this example:
> 
> * https://github.com/ets-labs/python-dependency-injector/blob/master/examples/miniapps/services_v1/run.py
> 
> For this dictionary from example:
> 
> ```python
> {'database': {'dsn': ':memory:'},
>                           'aws': {'access_key_id': 'KEY',
>                                   'secret_access_key': 'SECRET'},
>                           'auth': {'token_ttl': 3600}}
> ```
> 
> You should have some methods or lines that loads all configs before containers setup and final result convert to dict and paste it to `config.override(your_config)`

I think this is quite close to what I am doing. The only difference is that in the example he calls the override method from `__main__` while I call the override method from the `Core.__init__`. Is there any reason the former is better than the latter?

**Comment by rmk135** [comment]
Hi everybody,

Dependency Injector was designed to be a microframework, not an application framework. This is a reason why there is not enough builtin features for that. `Configuration` provider main responsibility is to implement late static binding - use configuration options before they are loaded. The topic of populating `Configuration` provider has not been covered intentionally.

The are numerous ways how configuration can be stored - ini / yaml files, environment variables, etc... Current design unifies all these to the dictionary that is passed to the `Configuration.override()`.

But things are going to change. In past few years I see constantly growing interest to the framework. People build applications from scratch with Dependency Injector. The lack of features for that becomes significant. I plan to cover the gap.

I re-open the issue as a feature.

Roman

**Comment by thegrymek** [comment]
Maybe leave this package as is and create separated repo this suffix `-plugins` for configuration features

**Comment by rmk135** [comment]
> Maybe leave this package as is and create separated repo this suffix `-plugins` for configuration features

That's not a bad idea.

**Comment by LastStarDust** [comment]
Sorry to bother you again.
What if I want to override only a single element or a subset of a configuration?
For example, what if I want to override only `option2` [here](http://python-dependency-injector.ets-labs.org/api/providers.html#dependency_injector.providers.Configuration)?

**Comment by rmk135** [comment]
@LastStarDust You can override individual value like that:

```python
config.section1.option2.override('new_value')
```

**Comment by LastStarDust** [comment]
Thank you very much

**Comment by rmk135** [comment]
@thegrymek @LastStarDust

Gentlemen,

The day has come. 

http://python-dependency-injector.ets-labs.org/providers/configuration.html#loading-from-ini-file

Available since version `3.18`.
